### PR TITLE
Add flitering of versions.

### PR DESCRIPTION
**Pull request by Darth-Ness**
Unfortunately,  I wasn't able to confirm it works. It does at least compile without errors. Let me know if you want me to change anything.

### CODE REVIEW

**Comment by tbrand** [comment]
As I wrote in https://github.com/negyio/negy/issues/13, you have to define `--min-version` arg frst.
Also, you shouldn't filtering the nodes by `while` ([here](https://github.com/negyio/negy/pull/18/files#diff-8c6588549035c43e116126c4af4ff11572b6f751a8dc99ddde44d726a5221047R62)) but you should do it in the iterator [here](https://github.com/negyio/negy/pull/18/files#diff-8c6588549035c43e116126c4af4ff11572b6f751a8dc99ddde44d726a5221047R57)

**Comment by Darth-Ness** [comment]
> As I wrote in #13, you have to define `--min-version` arg frst. Also, you shouldn't filtering the nodes by `while` ([here](https://github.com/negyio/negy/pull/18/files#diff-8c6588549035c43e116126c4af4ff11572b6f751a8dc99ddde44d726a5221047R62)) but you should do it in the iterator [here](https://github.com/negyio/negy/pull/18/files#diff-8c6588549035c43e116126c4af4ff11572b6f751a8dc99ddde44d726a5221047R57)

Okay, it now uses the filter method instead of a while loop.

**Comment by tbrand** [comment]
Cool!
- Avoid parsing args again [here](https://github.com/negyio/negy/pull/18/files#diff-8c6588549035c43e116126c4af4ff11572b6f751a8dc99ddde44d726a5221047R51) but pass it as arg.
- Is [this](https://github.com/negyio/negy/pull/18/files#diff-8c6588549035c43e116126c4af4ff11572b6f751a8dc99ddde44d726a5221047R61) correct? It seems the opposite.

Please create a latest docker-compose environment to confirm the implementation! Only you have to do is download the latest docker-compose binary [here](https://github.com/docker/compose/releases).

**Comment by Darth-Ness** [comment]
Do you mean the operator I used? If so I think it correct.

**Comment by Darth-Ness** [comment]
`sudo ./docker-compose-linux-armv7 build` outputs

>  ERROR [negy-negy-node1 internal] load metadata for docker.io/library/debian:buster-slim

**Comment by tbrand** [comment]
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter
`true` elements will be remaining. so it should be n.version >= min_version

File: negy-gateway/src/main.rs [hunk 1]
```diff
@@ -22,6 +22,8 @@ struct Args {
     node_pool_endpoint: String,
     #[clap(short, long, value_parser, default_value = "3")]
     hops: usize,
+    #[clap(short, long, value_parser, default_value = "0.1.5")]
+    min_version: String,
```
Comment (tbrand) [inline negy-gateway/src/main.rs]: could you make this arg option? i mean `min_version: Option<String>`

**Review by tbrand — CHANGES_REQUESTED** [review]

File: negy-gateway/src/main.rs (hunk 1 above)
Comment (Darth-Ness) [reply negy-gateway/src/main.rs]: I did that (on my local copy) and that outputted this error.
> can't compare `std::string::String` with `Option<std::string::String>

Comment (tbrand) [reply negy-gateway/src/main.rs]: Yup the error says what it is. You can't compare String and Option<String>.

Comment (Darth-Ness) [reply negy-gateway/src/main.rs]: So then I would need to convert the option back to string?

Comment (tbrand) [reply negy-gateway/src/main.rs]: You should do `If the min_version is None (not specified), every node is a target. If the min_version is specified, apply the filter`. It gonna be like this.

```
.filter(|n| {
    if let Some(min_version) = args.min_version {
        n.version >= min_version
    } else {
        true
    }
})
```

Comment (Darth-Ness) [reply negy-gateway/src/main.rs]: Like that?

File: negy-gateway/src/main.rs [hunk 2]
```diff
@@ -55,6 +58,14 @@ async fn fetch_nodes_unselected(node_pool_endpoint: &str) -> Result<Vec<NodeUnse
             name: n.name,
             version: n.version,
         })
+        .filter(|n| {
+           if let Some(min_version) = &args.min_version {
+                n.version.parse::<i32>().unwrap() >= min_version.parse::<i32>().unwrap()
```
Comment (tbrand) [inline negy-gateway/src/main.rs]: Don't parse as i32 but add `semver` crate and parse as Version. https://docs.rs/semver/latest/semver/

File: negy-gateway/src/main.rs [hunk 3]
```diff
@@ -46,6 +48,7 @@ async fn fetch_nodes_unselected(node_pool_endpoint: &str) -> Result<Vec<NodeUnse
         .await?
         .json::<ListNodeResponse>()
         .await?;
+    let args = Args::parse();
```
Comment (tbrand) [inline negy-gateway/src/main.rs]: Avoid `Args::parse` here but pass the min_version as an arg of `spawn_inner`. It's already parsed in main function.

**Review by tbrand — CHANGES_REQUESTED** [review]

File: negy-gateway/src/main.rs (hunk 3 above)
Comment (Darth-Ness) [reply negy-gateway/src/main.rs]: Okay, it now passes args from the main function.

**Comment by Darth-Ness** [comment]
Should I focus on resolving the conflict now?

**Comment by tbrand** [comment]
Sure!

**Comment by Darth-Ness** [comment]
Okay, I got the conflict resolved.

**Comment by tbrand** [comment]
Nice! The last thing you have to do it add semver create as I wrote.

File: negy-gateway/src/main.rs [hunk 4]
```diff
@@ -24,6 +24,8 @@ struct Args {
     hops: usize,
     #[clap(short, long, value_parser)]
     auth_token: Option<String>,
+    min_version: Option<String>,
+
```
Comment (tbrand) [inline negy-gateway/src/main.rs]: this is a meaningless empty line. please remove it.

File: negy-gateway/src/main.rs [hunk 5]
```diff
@@ -68,14 +76,16 @@ async fn spawn(
     node_pool_endpoint: String,
     hops: usize,
     auth_token: Option<String>,
+    min_version: Option<String>,
 ) -> Result<()> {
+
```
Comment (tbrand) [inline negy-gateway/src/main.rs]: this is a meaningless empty line. please remove it.

File: negy-gateway/src/main.rs [hunk 6]
```diff
@@ -124,11 +134,13 @@ async fn main() -> Result<()> {
 
     let listener = TcpListener::bind(bind_addr).await?;
 
+
```
Comment (tbrand) [inline negy-gateway/src/main.rs]: this is a meaningless empty line. please remove it.

**Comment by Darth-Ness** [comment]
Okay empty lines removed.

**Comment by tbrand** [comment]
Thxs. Please add semver crate as I mentioned.

**Comment by Darth-Ness** [comment]
All right, it now uses the semver crate.

File: negy-gateway/src/main.rs [hunk 7]
```diff
@@ -68,14 +77,16 @@ async fn spawn(
     node_pool_endpoint: String,
     hops: usize,
     auth_token: Option<String>,
+    min_version: Option<String>,
 ) -> Result<()> {
+
```
Comment (tbrand) [inline negy-gateway/src/main.rs]: empty line added

File: negy-gateway/src/main.rs [hunk 8]
```diff
@@ -124,11 +135,13 @@ async fn main() -> Result<()> {
 
     let listener = TcpListener::bind(bind_addr).await?;
 
+
```
Comment (tbrand) [inline negy-gateway/src/main.rs]: empty line added

File: negy-gateway/src/main.rs [hunk 9]
```diff
@@ -24,6 +25,8 @@ struct Args {
     hops: usize,
     #[clap(short, long, value_parser)]
     auth_token: Option<String>,
+    min_version: Option<String>,
+
```
Comment (tbrand) [inline negy-gateway/src/main.rs]: empty line added

**Comment by tbrand** [comment]
Looks good!! After you removing the empty lines, I gonna merge this!

**Comment by Darth-Ness** [comment]
Hang on, I forgot to update the cargo files, let me do that before you merge.

**Comment by Darth-Ness** [comment]
Okay got those updated, and empty lines removed.

**Comment by tbrand** [comment]
Yup. thanks

**Comment by tbrand** [comment]
I think you need to add semver both of /Cargo.toml and /negy-gateway/Cargo.toml. Please confirm your code could be compiled and run successfully. Also, please do not commit any changes which is meaningless in this pull request. Like adding empty lines and removing existing codes. Thanks.

**Comment by Darth-Ness** [comment]
I did fix that one file. Not sure why that happened. And yes, `cargo run` in the negy-gateway directory work just fine

**Comment by tbrand** [comment]
Hey @Darth-Ness, I would like to finish this PR for the next release so I created a new PR to fix this. https://github.com/negyio/negy/pull/22
All of your commits are in it. If it's ok, I would like to merge #22. How does it sound?

**Comment by Darth-Ness** [comment]
> Hey @Darth-Ness, I would like to finish this PR for the next release so I created a new PR to fix this. #22 All of your commits are in it. If it's ok, I would like to merge #22. How does it sound?

That's good, go ahead and merge it whenever your ready. Do you want me to close this?
By the way, when some one makes a PR to one of your repos, you can directly commit to the branch the new commits are coming from (Unless that is turned off)

**Comment by tbrand** [comment]
I merged this! Thanks for the contribution!! :tada: Your name will be listed.

**Pull request merged**